File: php/plugins/runcmd.php
<?php

/**
 * The child-process half of the two plugins that shell out - `boru` and
 * `secretspec` - in one place and OUTSIDE the core: a chain of built-ins
 * never requires this file, and so never reaches `proc_open`.
 */

declare(strict_types=1);

namespace Voxgig\Sekreto\Plugins;

require_once __DIR__ . '/../src/Sekreto.php';

use Voxgig\Sekreto\SekretoError;

/**
 * Run a child to completion and collect both its streams.
 *
 * Given an ARGV ARRAY, not a command string. `proc_open` with a string runs
 * `/bin/sh -c`, which put every value - the command, the alias, the key -
 * one escaping function away from being interpreted. `escapeshellarg` was
 * doing that job correctly, but it made this the only port whose safety
 * rested on quoting rather than on `execve`, and it broke two things
 * besides: the "cannot run" branch was unreachable, because `sh` always
 * starts and absorbs the exec failure into its own message, and the
 * `BORU_HOME=...` prefix is POSIX-sh syntax that does nothing on Windows,
 * where `proc_open` goes through `cmd.exe`. The array form (PHP 7.4+) makes
 * this port behave like the other eleven.
 *
 * Both pipes are drained TOGETHER. `stream_get_contents` on stdout to EOF
 * and only then on stderr deadlocks the moment the child writes more than
 * one pipe buffer (64 KiB on Linux) to stderr: this process is blocked
 * waiting for stdout, the child is blocked waiting for room on stderr, and
 * neither can move. Nothing in this library sets a timeout, so that hang is
 * permanent. secretspec's diagnostics are box-drawn and reach that size
 * easily.
 *
 * The child's stdin is closed rather than inherited, so a CLI that reads it
 * - one prompting for a passphrase when its environment variable is absent -
 * sees EOF and gives up instead of waiting forever.
 *
 * @param list<string>          $argv
 * @param array<string, string> $env  extra variables for the child only
 *
 * @return array{0: string, 1: string, 2: int} stdout, trimmed stderr, status
 */
function runcmd(array $argv, array $env = []): array
{
    $descriptors = [0 => ['pipe', 'r'], 1 => ['pipe', 'w'], 2 => ['pipe', 'w']];

    // null inherits the parent environment; a non-empty $env has to be the
    // whole set, so it is merged rather than replacing it.
    $childenv = [] === $env ? null : array_merge(getenv(), $env);

    $process = @proc_open($argv, $descriptors, $pipes, null, $childenv);

    if (!is_resource($process)) {
        throw new SekretoError('sekreto: cannot run ' . $argv[0]);
    }

    fclose($pipes[0]);

    stream_set_blocking($pipes[1], false);
    stream_set_blocking($pipes[2], false);

    $text = [1 => '', 2 => ''];
    $open = [1 => $pipes[1], 2 => $pipes[2]];

    while ([] !== $open) {
        $read = array_values($open);
        $write = null;
        $except = null;

        if (false === @stream_select($read, $write, $except, 30)) {
            break;
        }

        foreach ($read as $handle) {
            $chunk = fread($handle, 65536);
            $key = $handle === $pipes[1] ? 1 : 2;

            if (false === $chunk || ('' === $chunk && feof($handle))) {
                unset($open[$key]);
                continue;
            }

            $text[$key] .= $chunk;
        }
    }

    fclose($pipes[1]);
    fclose($pipes[2]);

    $status = proc_close($process);

    return [$text[1], trim($text[2]), $status];
}

File: ocaml/plugins/runcmd.ml
(* Running a store's own CLI, and collecting both its streams.

   Two of the ten plugin kinds read their store through a child process -
   `boru` and `secretspec` - and a child process is one of the three things
   that keeps a kind out of the core, beside a socket and a signature. *)

open Secret

type ran = { out : string; why : string; status : int }

(* Where a command lives, or nowhere.

   Resolved before forking: OCaml's `create_process` execs in the child, so
   a command that is not there makes the child exit 127 with no error in the
   parent - which is indistinguishable from a command that ran and failed.
   "cannot run" and "the store said no" are different answers. *)
let findexe (command : string) : string option =
  let executable path = match Unix.access path [ Unix.X_OK ] with () -> true | exception _ -> false in

  if String.contains command '/' then if executable command then Some command else None
  else
    let path = match Sys.getenv_opt "PATH" with Some value -> value | None -> "/usr/bin:/bin" in
    let rec walk = function
      | [] -> None
      | dir :: rest ->
        let cand = (if "" = dir then "." else dir) ^ "/" ^ command in
        if executable cand then Some cand else walk rest
    in
    walk (String.split_on_char ':' path)

(* Run a child to completion and collect both its streams.

   The two streams are drained CONCURRENTLY, through one select loop.
   Reading stdout to EOF and only then reading stderr deadlocks the moment
   the child writes more than one pipe buffer (64 KiB on Linux) to stderr:
   the parent is blocked waiting for stdout, the child is blocked waiting
   for room on stderr, and neither can move. Nothing here sets a timeout, so
   that hang is permanent - `get` simply never returns. secretspec's
   diagnostics are box-drawn and reach that size easily.

   The child's stdin is closed rather than left open on a pipe nobody writes
   to, so a CLI that reads it - one prompting for a passphrase when its
   environment variable is absent - sees EOF and gives up instead of waiting
   forever.

   Arguments go as an array, never through a shell, and no secret is ever
   put on a command line where the process table would publish it. *)
let runcmd (argv : string list) (env : string array) (command : string) : ran =
  let exe =
    match findexe command with
    | Some path -> path
    | None -> fail ("sekreto: cannot run " ^ command ^ ": No such file or directory")
  in

  let outr, outw = Unix.pipe () in
  let errr, errw = Unix.pipe () in
  let inr, inw = Unix.pipe () in
  Unix.close inw;

  let pid =
    match Unix.create_process_env exe (Array.of_list argv) env inr outw errw with
    | pid -> pid
    | exception Unix.Unix_error (err, _, _) ->
      List.iter (fun fd -> try Unix.close fd with _ -> ()) [ inr; outr; outw; errr; errw ];
      fail ("sekreto: cannot run " ^ command ^ ": " ^ Unix.error_message err)
  in

  Unix.close inr;
  Unix.close outw;
  Unix.close errw;

  let obuf = Buffer.create 4096 and ebuf = Buffer.create 4096 in
  let live = ref [ outr; errr ] in
  let chunk = Bytes.create 65536 in

  while [] <> !live do
    let ready, _, _ = Unix.select !live [] [] (-1.0) in
    List.iter
      (fun fd ->
        let got = try Unix.read fd chunk 0 65536 with Unix.Unix_error _ -> 0 in
        if 0 = got then begin
          (try Unix.close fd with _ -> ());
          live := List.filter (fun other -> other <> fd) !live
        end
        else Buffer.add_subbytes (if fd = outr then obuf else ebuf) chunk 0 got)
      ready
  done;

  let status = match snd (Unix.waitpid [] pid) with
    | Unix.WEXITED code -> code
    | Unix.WSIGNALED code | Unix.WSTOPPED code -> 128 + code
  in

  { out = Buffer.contents obuf; why = trim (Buffer.contents ebuf); status }
